File: py/yeriasdk/views/message_view.py
"""
MessageView - A view for displaying messages to users
"""

from typing import Literal, Optional, Dict, Any, Union
from datetime import datetime

from ..core.base_view import BaseView
from ..types.models import SubmitAction, HttpMethod
from ..errors.exceptions import InvalidParameterError

MessageSeverity = Literal["info", "success", "warning", "error"]


class MessageView(BaseView):
    """Builds a Message SGUI view — a titled notice/dialog with a severity and
    up to two actions.

    ``set_body`` / ``set_intro`` / ``set_severity`` set the content and tone;
    ``set_primary_action`` and ``set_secondary_action`` (aliased by
    ``submit_button``) define the confirm/cancel buttons, and
    ``set_dismissible`` controls whether it can be closed without acting.

    Extends ``BaseView``; instantiated by the YeriaApp/YeriaUI factory,
    populated with these builders, then serialized to a JSON view description
    and signed into a v3 envelope by ``serve()``.
    """
    @classmethod
    def from_json(cls, json_view):
        """Rehydrate a Message view from a wire JSON payload."""
        return cls.from_json_as('Message', json_view)


    def __init__(self, view_id: str, title: str, process_id: Optional[str] = None):
        super().__init__(
            {
                "id": view_id,
                "type": "Message",
                "process_id": process_id,
                "metadata": {
                    "version": "1.0.0",
                    "created_at": datetime.now(),
                },
            }
        )

        self.content = {
            "title": title,
            "body": "",
            "severity": "info",
            "actions": [],
        }

    def set_intro(self, intro: str) -> "MessageView":
        """Set introduction text"""
        return self._set_intro_text("intro", intro)

    def set_body(self, body: str) -> "MessageView":
        """Set main message body"""
        if not body or not body.strip():
            raise InvalidParameterError("body", body, "Message body cannot be empty")

        self.content["body"] = body.strip()
        return self

    def set_severity(self, severity: MessageSeverity) -> "MessageView":
        """Set message severity"""
        self.content["severity"] = severity
        return self

    def add_action(
        self,
        label: str,
        go: Optional[str] = None,
        method: Optional[HttpMethod] = None,
        back: Optional[Union[int, str]] = None,
    ) -> "MessageView":
        """Add a button to the box. Two at most.

        Sans ``go``, il FERME et ne fait rien d'autre. Avec ``go``, il charge
        cette vue, qui entre dans la pile selon ses propres regles.

        Sans aucune action, le client dessine un unique « OK » qui ferme : une
        boite a toujours une sortie, comme une MsgBox en avait toujours une.

        Un identifiant de vue nu est refuse, comme partout ailleurs : rien sur
        le fil ne le distingue d'un chemin relatif.
        """
        trimmed = (label or "").strip()
        if not trimmed:
            raise InvalidParameterError(
                "label", label, "Action label cannot be empty"
            )

        actions = self.content["actions"]
        if len(actions) >= 2:
            raise InvalidParameterError(
                "actions", len(actions) + 1, "A message carries two actions at most"
            )

        if go is not None and back is not None:
            raise InvalidParameterError(
                "back",
                back,
                "an action either loads a view (`go`) or steps back (`back`), not both",
            )

        action: Dict[str, Any] = {"label": trimmed}
        if back is not None:
            # Reculer ne recharge RIEN : l'ecran vise est deja dans la pile,
            # avec l'etat ou l'utilisateur l'avait laisse. 'root' y revient
            # sans compter — c'est la seule position qu'un fournisseur
            # connaisse a coup sur, la vue servie par son URL de base.
            if back != "root":
                if isinstance(back, bool) or not isinstance(back, int) or back < 1:
                    raise InvalidParameterError(
                        "back", back, "back must be an integer >= 1, or 'root'"
                    )
            action["back"] = back
            if method is not None:
                raise InvalidParameterError(
                    "method", method, "method only applies to an action that carries `go`"
                )
        elif go is not None:
            action["go"] = self._assert_addressable_target("go", go)
            if method is not None:
                action["method"] = method
        elif method is not None:
            raise InvalidParameterError(
                "method", method, "method only applies to an action that carries `go`"
            )

        actions.append(action)
        return self

    def clear_actions(self) -> "MessageView":
        """Remove every action."""
        self.content["actions"] = []
        return self

    def set_primary_action(
        self, text: str, method: HttpMethod = "POST", confirm_message: Optional[str] = None
    ) -> "MessageView":
        """Nom historique.

        La destination voyageait dans ``confirm_message``, un champ nomme pour
        un texte de confirmation — c'est ``go`` desormais.
        """
        if confirm_message is None:
            return self.add_action(text)
        return self.add_action(text, go=confirm_message, method=method)

    def submit_button(
        self, text: str, method: HttpMethod = "POST", confirm_message: Optional[str] = None
    ) -> "MessageView":
        """Nom historique de add_action."""
        return self.set_primary_action(text, method, confirm_message)

    def set_secondary_action(
        self, text: str, method: HttpMethod = "POST", confirm_message: Optional[str] = None
    ) -> "MessageView":
        """Nom historique : la seconde action."""
        return self.set_primary_action(text, method, confirm_message)

    def clear_secondary_action(self) -> "MessageView":
        """Retire la seconde action, si elle existe."""
        del self.content["actions"][1:]
        return self

    def set_metadata(self, metadata: Dict[str, Any]) -> "MessageView":
        """Add message-specific metadata"""
        self.content["meta"] = dict(metadata)
        return self

File: js/src/core/message-view.ts
import { BaseView } from './base-view';
import { HttpMethod } from '../types';
import { InvalidParameterError } from '../errors';

export type MessageSeverity = 'info' | 'success' | 'warning' | 'error';

/**
 * Un bouton de la boîte, et ce qu'il fait.
 *
 * Sans `go`, il FERME — c'est tout ce qu'il fait. Avec `go`, il charge cette
 * vue, qui entre dans la pile selon ses propres règles.
 *
 * Le modèle vient des boîtes de dialogue classiques : un jeu de boutons, une
 * icône de gravité, et une issue par bouton. La seule différence tient à la
 * distance : là où `MsgBox` rendait à l'appelant le bouton pressé pour qu'il
 * décide, le fournisseur n'est pas là au moment du clic — l'issue voyage donc
 * avec le bouton.
 */
export interface MessageAction {
    label: string;
    go?: string;            // chemin ou URL : charge cette vue
    method?: HttpMethod;    // avec `go` seulement ; défaut GET
    back?: number | 'root'; // recule dans la pile, sans rien recharger
    // Aucun des deux ⇒ le bouton ferme, et c'est tout.
}

export interface MessageContent {
    title: string;
    intro?: string;
    body: string;
    severity: MessageSeverity;
    /** Zéro, une ou deux. Vide ⇒ le client dessine un unique « OK ». */
    actions: MessageAction[];
    meta?: Record<string, unknown>;
}

/**
 * Builds a Message SGUI view — a titled notice/dialog with a severity and up to
 * two actions.
 *
 * `setBody`/`setIntro`/`setSeverity` set the content and tone; `setPrimaryAction`
 * and `setSecondaryAction` (aliased by `submitButton`) define the confirm/cancel
 * buttons, and `setDismissible` controls whether it can be closed without acting.
 *
 * Extends {@link BaseView}; instantiated by the YeriaApp/YeriaUI factory,
 * populated with these builders, then serialized to a JSON view description and
 * signed into a v3 envelope by `serve()`.
 */
export class MessageView extends BaseView {

    static fromJson(json: Record<string, unknown>): MessageView {
        return MessageView.fromJsonAs(MessageView, 'Message', json);
    }
    constructor(viewId: string, title: string, processId?: string) {
        super({
            id: viewId,
            type: 'Message',
            processId,
            metadata: {
                version: '1.0.0',
                createdAt: new Date()
            }
        });

        this.content = {
            title,
            body: '',
            severity: 'info',
            actions: []
        } as MessageContent;
    }

    /**
     * Sets the introduction text
     */
    setIntro(intro: string): this {
        return this.setIntroText('intro', intro);
    }

    /**
     * Sets the main body of the message
     */
    setBody(body: string): this {
        if (!body || body.trim().length === 0) {
            throw new InvalidParameterError('body', body, 'Message body cannot be empty');
        }

        (this.content as MessageContent).body = body.trim();
        return this;
    }

    /**
     * Sets the message severity (info, success, warning, error)
     */
    setSeverity(severity: MessageSeverity): this {
        (this.content as MessageContent).severity = severity;
        return this;
    }

    /**
     * Adds a button to the box. Two at most.
     *
     * Without `go`, it closes and does nothing else. With `go`, it loads that
     * view — which then enters the stack under its own rules, like any other
     * navigation.
     *
     * With no action at all, the client draws a single « OK » that closes:
     * a box always has a way out, as a `MsgBox` always had one.
     *
     * A bare view id is refused, as everywhere else: nothing on the wire
     * distinguishes it from a relative path, and no client keeps a directory
     * of view ids.
     */
    addAction(
        label: string,
        options: { go?: string; method?: HttpMethod; back?: number | 'root' } = {}
    ): this {
        const trimmed = (label ?? '').trim();
        if (trimmed.length === 0) {
            throw new InvalidParameterError('label', label, 'Action label cannot be empty');
        }

        const actions = (this.content as MessageContent).actions;
        if (actions.length >= 2) {
            throw new InvalidParameterError(
                'actions',
                actions.length + 1,
                'A message carries two actions at most'
            );
        }

        if (options.go !== undefined && options.back !== undefined) {
            throw new InvalidParameterError(
                'back',
                options.back,
                'an action either loads a view (`go`) or steps back (`back`), not both'
            );
        }

        const action: MessageAction = { label: trimmed };
        if (options.back !== undefined) {
            // Reculer ne recharge RIEN : l'écran visé est déjà dans la pile,
            // avec l'état où l'utilisateur l'avait laissé. `'root'` y revient
            // sans compter — c'est la seule position qu'un fournisseur
            // connaisse à coup sûr, la vue servie par son URL de base.
            if (options.back !== 'root') {
                if (!Number.isInteger(options.back) || options.back < 1) {
                    throw new InvalidParameterError(
                        'back',
                        options.back,
                        "back must be an integer >= 1, or 'root'"
                    );
                }
            }
            action.back = options.back;
            if (options.method !== undefined) {
                throw new InvalidParameterError(
                    'method',
                    options.method,
                    'method only applies to an action that carries `go`'
                );
            }
        } else if (options.go !== undefined) {
            action.go = this.assertAddressableTarget('go', options.go);
            if (options.method !== undefined) action.method = options.method;
        } else if (options.method !== undefined) {
            throw new InvalidParameterError(
                'method',
                options.method,
                'method only applies to an action that carries `go`'
            );
        }

        actions.push(action);
        return this;
    }

    /** Removes every action. */
    clearActions(): this {
        (this.content as MessageContent).actions = [];
        return this;
    }

    /**
     * Nom historique. La destination voyageait dans `confirmMessage`, un champ
     * nommé pour un texte de confirmation — c'est `go` désormais.
     */
    setPrimaryAction(text: string, method: HttpMethod = 'POST', confirmMessage?: string): this {
        return this.addAction(text, confirmMessage === undefined
            ? {}
            : { go: confirmMessage, method });
    }

    /** Nom historique de {@link addAction}. */
    submitButton(text: string, method: HttpMethod = 'POST', confirmMessage?: string): this {
        return this.setPrimaryAction(text, method, confirmMessage);
    }

    /** Nom historique : la seconde action. */
    setSecondaryAction(text: string, method: HttpMethod = 'POST', confirmMessage?: string): this {
        return this.setPrimaryAction(text, method, confirmMessage);
    }

    /** Retire la seconde action, si elle existe. */
    clearSecondaryAction(): this {
        const actions = (this.content as MessageContent).actions;
        if (actions.length > 1) actions.length = 1;
        return this;
    }

    /**
     * Adds message-specific metadata
     */
    setMetadata(metadata: Record<string, unknown>): this {
        (this.content as MessageContent).meta = { ...metadata };
        return this;
    }

}
